### PR TITLE
Support multi-GPU validation outside of training

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22568 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Add native support and docs for multi-GPU validation using torch.distributed.run, enabling DDP-aware dataloading across tasks. ⚡

### 📊 Key Changes
- Docs: New “Multi-GPU Validation” section with a ready-to-run script and launch command example using `torch.distributed.run`.
- Detect: Use `rank=RANK` in validator DataLoader construction instead of `rank=-1`.
- Classify: Use `rank=RANK` in validator DataLoader construction instead of `rank=-1`.

### 🎯 Purpose & Impact
- Enable out-of-the-box multi-GPU validation with proper sampler sharding and device assignment.
- Improve scalability and throughput for large datasets, aligning validation UX with distributed training.
- Maintain backward compatibility for single-GPU/CPU runs; no device arg needed when using DDP.
- Clear guidance for users to launch multi-GPU validation with minimal setup. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review helper; an Ultralytics engineer will follow up shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The new multi-GPU validation docs still need backend initialization for the process group and should configure the CUDA device with the local rank so the example works beyond single-node runs.

💬 Posted 2 inline comments

File: docs/en/modes/val.md [hunk 1]
```diff
@@ -142,6 +142,34 @@ The below examples showcase YOLO model validation with custom arguments in Pytho
...
+```python
+# Save as multi_gpu_val.py
+import torch
+import torch.distributed as dist
+from ultralytics import YOLO
+from ultralytics.utils import RANK
+
+torch.cuda.set_device(RANK)
+dist.init_process_group()
+
```
Comment (Y-T-G) [inline docs/en/modes/val.md]: It's optional https://docs.pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group

File: docs/en/modes/val.md [hunk 2]
```diff
@@ -142,6 +142,35 @@ The below examples showcase YOLO model validation with custom arguments in Pytho
...
+```python
+# Save as multi_gpu_val.py
+import torch
+import torch.distributed as dist
+
+from ultralytics import YOLO
+from ultralytics.utils import LOCAL_RANK
+
+torch.cuda.set_device(LOCAL_RANK)
+dist.init_process_group()
```
Comment (UltralyticsAssistant) [inline docs/en/modes/val.md]: 💡 **MEDIUM**: `dist.init_process_group()` requires a backend argument (for example, `"nccl"` for multi-GPU CUDA runs). Calling it without a backend raises a `TypeError`, so the example as written will fail. Please add the backend parameter to make the snippet runnable.

**Suggested change:**
```suggestion
dist.init_process_group("nccl")
```

File: ultralytics/models/yolo/classify/val.py [hunk 3]
```diff
@@ -167,7 +167,7 @@ def get_dataloader(self, dataset_path: Path | str, batch_size: int) -> torch.uti
             (torch.utils.data.DataLoader): DataLoader object for the classification validation dataset.
         """
         dataset = self.build_dataset(dataset_path)
-        return build_dataloader(dataset, batch_size, self.args.workers, rank=-1)
+        return build_dataloader(dataset, batch_size, self.args.workers, rank=RANK)
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/classify/val.py]: ⚠️ **HIGH**: Using `rank=RANK` here causes the validation DataLoader to shard the dataset when the trainer is running under DDP, but we still execute validation only on rank 0. As a result, rank 0 only sees a 1/world_size slice of the data and the reported metrics are incorrect. We should keep `rank=-1` (or gate the sharding behind a new flag) unless we also execute validation on every rank.

File: ultralytics/models/yolo/detect/val.py [hunk 4]
```diff
@@ -316,7 +316,7 @@ def get_dataloader(self, dataset_path: str, batch_size: int) -> torch.utils.data
             batch_size,
             self.args.workers,
             shuffle=False,
-            rank=-1,
+            rank=RANK,
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/val.py]: ⚠️ **HIGH**: With this change the validation DataLoader uses a distributed sampler whenever DDP is active, but the validator still runs only on rank 0 during training. That combination means validation now processes only 1/world_size of the images, giving incorrect metrics. We should avoid enabling distributed sharding here unless validation truly runs on every rank.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Block merging until we prevent the validators from sharding the dataset on rank 0-only runs and fix the distributed init example to include the required backend argument.

💬 Posted 3 inline comments

**Comment by glenn-jocher** [comment]
@Y-T-G ideally we want to abstract away all the DDP complication as we've done for training. This would be taking things a step back in terms of user experience. 

Maybe we could refactor our DDP code to support both training and validation?

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please restore the correct model assignment in the generated DDP helper so the subprocess receives a valid string path, and guard the new dataloader batch split against `world_size == 0` to keep CPU validation working.

💬 Posted 2 inline comments

File: ultralytics/engine/validator.py [hunk 5]
```diff
@@ -153,37 +161,45 @@ def __call__(self, trainer=None, model=None):
             self.args.plots &= trainer.stopper.possible_stop or (trainer.epoch == trainer.epochs - 1)
             model.eval()
         else:
+            self.world_size = get_world_size(self.args.device)
+            if self.world_size > 1 and RANK == -1:
+                self.model = model
+                run_ddp(self)
+                return
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/validator.py:168]: ⚠️ **HIGH**: Spawning the distributed subprocess and then returning here means the caller never receives the validation metrics. For example, `metrics = model.val(device=[0,1])` now yields `None`, whereas the single-GPU path returns the metrics dict. We need to collect the subprocess results (e.g. share via a temp file or dist broadcast) and return them so the Python API keeps working.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Distributed validation launches correctly, but the parent process exits early without returning metrics. Please wire the subprocess results back so multi-GPU validation keeps the same Python API contract as before.

💬 Posted 1 inline comment

File: ultralytics/engine/validator.py (hunk 5 above)
Comment (Y-T-G) [reply ultralytics/engine/validator.py:168]: We never return metrics for multi-GPU

**Comment by Y-T-G** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2025-11-04 18:52:27 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/19079547329): `ci.yml` run 14049
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/19079547561): `docker.yml` run 1378

**Comment by Y-T-G** [comment]
@glenn-jocher I refactored the DDP code to for reuse to support both multi-GPU training multi-GPU validation through the Python API and CLI. It works from my testing.

**Comment by Y-T-G** [comment]
There's a parallel PR by @baibaidj 

https://github.com/ultralytics/ultralytics/pull/22579

There are some changes there that are useful like labels dir guard but I didn't incorporate them here since that doesn't add him as a contributor, even if I use the co-author feature. I guess we can refactor his PR to keep the fixes and then merge that later.

**Comment by Laughing-q** [comment]
@Y-T-G Nice! I wanted to do the other day and wanted to make a PR but got stuck on combing DDP code with our Autobackend wrapper so gave up.

**Comment by baibaidj** [comment]
> There's a parallel PR by @baibaidj
> 
> #22579
> 
> There are some changes there that are useful like labels dir guard but I didn't incorporate them here since that doesn't add him as a contributor, even if I use the co-author feature. I guess we can refactor his PR to keep the fixes and then merge that later.

Nice work! It is great to see that my implementation is not so far away from yours and I'm happy that some of my code has been incorporated, Thanks.

**Comment by glenn-jocher** [comment]
Thanks all—direction looks good and keeps validation UX consistent with training. To land this cleanly: ensure validators pass `rank=RANK` to their dataloaders and rely on `gather_stats()` to merge metrics, which aligns with the validator flow and contiguous sampler behavior documented in [Detection Validator gather_stats](https://docs.ultralytics.com/reference/models/yolo/detect/val/#gather_stats), [build_dataloader](https://docs.ultralytics.com/reference/data/build/#ultralytics.data.build.build_dataloader) and the validator’s [`__call__`](https://docs.ultralytics.com/reference/engine/validator/#call). For the launcher, please route multi‑GPU `val` through our existing helpers instead of bespoke calls so users don’t touch `torch.distributed` flags—see [generate_ddp_command](https://docs.ultralytics.com/reference/utils/dist/#ultralytics.utils.dist.generate_ddp_command), [generate_ddp_file](https://docs.ultralytics.com/reference/utils/dist/#ultralytics.utils.dist.generate_ddp_file) and [ddp_cleanup](https://docs.ultralytics.com/reference/utils/dist/#ultralytics.utils.dist.ddp_cleanup). Let’s keep the labels‑dir guard in #22579 for attribution and scope, avoid any `default.yaml` changes here, add minimal tests for sampler selection and metric gathering, and tuck the DDP note into the [Model Validation docs](https://docs.ultralytics.com/modes/val/).

**Comment by Y-T-G** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2025-11-18 16:47:21 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/19473960565): `ci.yml` run 14521
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/19473960592): `docker.yml` run 1423
